File: OOP/6_multilevel_inheritance.cpp
#include <bits/stdc++.h>
using namespace std;

// Base class Person
class Person
{
public:
    string name;
    int age;

    // Constructor for Person
    Person(string name, int age)
    {
        this->name = name;
        this->age = age;
    }

    // Method to display personal info
    void personalInfo()
    {
        cout << name << " - " << age << endl;
    }
};

// Derived class Company from Person
class Company : public Person
{
public:
    string cname; // Company name
    string cloc;  // Company location

    // Constructor for Company
    Company(string name, int age, string cname, string cloc)
        : Person(name, age)
    { // Call the Person constructor
        this->cname = cname;
        this->cloc = cloc;
    }

    // Method to display company info
    void companyInfo()
    {
        cout << cname << " - " << cloc << endl;
    }
};

// Derived class Employee from Company
class Employee : public Company
{
public:
    // Constructor for Employee
    Employee(string name, int age, string cname, string cloc)
        : Company(name, age, cname, cloc)
    {
    }
};

int main()
{

    Employee obj("Sayed", 24, "Google", "USA");

    obj.personalInfo();
    obj.companyInfo();

    return 0;
}

File: OOP/6_multilevel_inheritance.py
class Person:
    def __init__(self, name, age) -> None:
        self.name = name
        self.age = age

    def personalInfo(self):
        print(self.name, " - ", self.age)


class Company(Person):
    def __init__(self, name, age, cname, cloc) -> None:
        # Call the Person constructor
        Person.__init__(self, name, age)
        self.cloc = cloc
        self.cname = cname

    def companyInfo(self):
        print(self.cname, " - ", self.cloc)


class Employee(Company):
    def __init__(self, name, age, cname, cloc) -> None:
        # Call the Company constructor
        Company.__init__(self, name, age, cname, cloc)


# Creating an object of Employee
obj = Employee("Sayed", 24, "Google", "USA")

# Display personal and company info
obj.personalInfo()
obj.companyInfo()

# Method resolution order
print(Employee.mro())  # This prints the method resolution order
